### PR TITLE
[Binapi] introduce NWI binary API for UPG plugin

### PR DESCRIPTION
**Pull request by sergeymatov**

### CODE REVIEW

File: test/e2e/upg_e2e.go [hunk 1]
```diff
@@ -476,6 +476,57 @@ var _ = ginkgo.Describe("Binapi", func() {
 			gomega.Expect(err).NotTo(gomega.BeNil())
 		})
 	})
+	ginkgo.Context("NWI binapi", func() {
```
Comment (ivan4th) [inline test/e2e/upg_e2e.go:479]: `"for NWIs"`

File: test/e2e/upg_e2e.go [hunk 2]
```diff
@@ -476,6 +476,57 @@ var _ = ginkgo.Describe("Binapi", func() {
...
+		f := framework.NewDefaultFramework(framework.UPGModeTDF, framework.UPGIPModeV4)
+		ginkgo.It("Check nwi binapi functions", func() {
```
Comment (ivan4th) [inline test/e2e/upg_e2e.go:481]: `"adds, removes and lists the NWIs"`

**Review by ivan4th — CHANGES_REQUESTED** [review]

File: test/e2e/upg_e2e.go [hunk 3]
```diff
@@ -476,6 +476,57 @@ var _ = ginkgo.Describe("Binapi", func() {
...
+				if stop {
+					break
+				}
+				if msg.Name != "testing" {
+					continue
+				}
+				found = true
+			}
+			gomega.Expect(found).NotTo(gomega.BeTrue())
+
```
[the remark this answers is not in the document]
Comment (sergeymatov) [reply test/e2e/upg_e2e.go]: done

File: upf/upf.c [hunk 4]
```diff
@@ -832,6 +832,43 @@ vnet_upf_policy_fn (fib_route_path_t * rpaths, u8 * policy_id, u8 action)
   return rc;
 }
 
+/**
+ * Translate "foo.com" into "0x3 f o o 0x3 c o m 0x0"
+ * A historical / hysterical micro-TLV scheme. DGMS.
+ */
```
[the remark this answers is not in the document]
Comment (sergeymatov) [reply upf/upf.c:840]: done

File: upf/upf.api [hunk 5]
```diff
@@ -215,3 +215,24 @@ define upf_policy_dump {
   u32 client_index;
   u32 context;
 };
+
+autoreply define upf_nwi_add_del {
+  u32 client_index;
+  u32 context;
+  u8 add;
+  string name[64];
+  u32 vrf;
```
[the remark this answers is not in the document]
Comment (sergeymatov) [reply upf/upf.api]: Adjusted api message structure with `ip[46]_table_id` and some logic around it inside UPG

File: test/e2e/upg_e2e.go [hunk 6]
```diff
@@ -431,7 +431,7 @@ func describePDRReplacement(f *framework.Framework) {
 var _ = ginkgo.Describe("Binapi", func() {
 	ginkgo.Context("policy based routing", func() {
 		f := framework.NewDefaultFramework(framework.UPGModeTDF, framework.UPGIPModeV4)
-		ginkgo.It("Check binapi functions", func() {
+		ginkgo.It("Check policy binapi functions", func() {
```
[the remark this answers is not in the document]
Comment (sergeymatov) [reply test/e2e/upg_e2e.go]: thanks, fixed

File: test/e2e/upg_e2e.go (hunk 1 above)
Comment (sergeymatov) [reply test/e2e/upg_e2e.go]: fixed

File: test/e2e/upg_e2e.go (hunk 2 above)
Comment (sergeymatov) [reply test/e2e/upg_e2e.go]: done

**Review by ivan4th — APPROVED** [review]
LGTM

**Pull request merged**